### PR TITLE
Feature/schema registry kafka reporter

### PR DESCRIPTION
**Pull request by KevinJMao**
Add support for Confluent's [Schema Registry](https://github.com/confluentinc/schema-registry) to the Kafka reporting backend.

### CODE REVIEW

File: src/main/java/com/verisign/storm/metrics/GraphiteMetricsConsumer.java [hunk 1]
```diff
@@ -202,6 +237,33 @@ else if (dataPoint.value instanceof Number) {
...
+  private Double convertToDoubleValue(String key, Object value) {
+    if (value != null) {
+      Double result;
+
+      if (value instanceof String) {
+        try {
+          result = Double.parseDouble((String) value);
+        }
+        catch (NumberFormatException e) {
+          LOG.warn("Unparseable metric with key {} and value of {} found. Discarding metric.", key, value);
```
Comment (melodious) [inline src/main/java/com/verisign/storm/metrics/GraphiteMetricsConsumer.java]: this message could be more clear and say specifically that it is trying to parse it into a Double

File: src/main/java/com/verisign/storm/metrics/GraphiteMetricsConsumer.java [hunk 2]
```diff
@@ -202,6 +237,33 @@ else if (dataPoint.value instanceof Number) {
...
+        result = ((Number) value).doubleValue();
+      }
+      else {
+        result = null;
+      }
+
+      return result;
+    }
+    else {
+      LOG.warn("Metric with key {} has a value of null. Discarding metric.", key);
```
Comment (melodious) [inline src/main/java/com/verisign/storm/metrics/GraphiteMetricsConsumer.java]: Is it actually possible to have a key of null?

File: src/test/java/com/verisign/storm/metrics/GraphiteMetricsConsumerTest.java [hunk 3]
```diff
@@ -41,26 +49,48 @@
...
+  }
+
   @Test public void shouldInitializeGraphiteReporter() {
     // Given a Graphite configuration and topology context
     Map<String, String> stormConfig = Maps.newHashMap();
     stormConfig.put("metrics.graphite.host", testGraphiteHost);
-    stormConfig.put(GraphiteMetricsConsumer.REPORTER_NAME, testReporterName);
-
+    stormConfig
+        .put(GraphiteMetricsConsumer.REPORTER_NAME, "com.verisign.storm.metrics.reporters.graphite.GraphiteReporter");
```
Comment (melodious) [inline src/test/java/com/verisign/storm/metrics/GraphiteMetricsConsumerTest.java]: you could probably something like this and your test will be more refactorable:
```
GraphiteReporter.class.getName()
```

File: src/test/java/com/verisign/storm/metrics/GraphiteMetricsConsumerTest.java [hunk 4]
```diff
@@ -151,57 +157,42 @@ public void shouldReadDataAndSendLong(Collection<DataPoint> dataPoints) {
...
-  @Test(dependsOnMethods = { "shouldInitializeGraphiteReporter" }, dataProvider = "generateDataPointsFloat")
-  public void shouldReadDataAndSendFloat(Collection<DataPoint> dataPoints) {
-    //Given a consumer
-    GraphiteMetricsConsumer consumer = spy(new GraphiteMetricsConsumer());
+  @DataProvider(name = "generateMapDataPoints") public Object[][] generateMapDataPoints() {
+    int numDataPoints = 5;
+    int numDataPointValues = 5;
 
-    Map stormConf = ImmutableMap
-        .of("metrics.graphite.host", testStormSrcWorkerHost, "metrics.graphite.port", testStormSrcWorkerPort.toString(),
```
Comment (melodious) [inline src/test/java/com/verisign/storm/metrics/GraphiteMetricsConsumerTest.java]: a constant or variable for all of these strings will make your tests more dynamic

File: src/test/java/com/verisign/storm/metrics/reporters/BaseKafkaReporterTest.java [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+    kafkaServer.shutdown();
+    try {
+      zookeeper.close();
+    }
+    catch (IOException e) {
+      LOG.error("Failed to close Zookeeper server at {}:{}", testZkHost, testZkPort);
+    }
+  }
+
+  @DataProvider(name = "metrics") public Object[][] metricsProvider() {
```
Comment (melodious) [inline src/test/java/com/verisign/storm/metrics/reporters/BaseKafkaReporterTest.java]: was the intent to have more than one Data Provider?

File: src/main/java/com/verisign/storm/metrics/GraphiteMetricsConsumer.java (hunk 1 above)
Comment (KevinJMao) [inline src/main/java/com/verisign/storm/metrics/GraphiteMetricsConsumer.java]: Changed to 
```java
LOG.warn("Failed to parse metric with key {} and value of {} into a Double. Discarding metric.", key, value);
```

File: src/test/java/com/verisign/storm/metrics/GraphiteMetricsConsumerTest.java (hunk 3 above)
Comment (KevinJMao) [inline src/test/java/com/verisign/storm/metrics/GraphiteMetricsConsumerTest.java]: Replaced hardcoded class names with `<ReporterClass>.class.getName()`

File: src/test/java/com/verisign/storm/metrics/GraphiteMetricsConsumerTest.java (hunk 4 above)
Comment (KevinJMao) [inline src/test/java/com/verisign/storm/metrics/GraphiteMetricsConsumerTest.java]: Replaced hardcoded field names with static constants.

File: src/test/java/com/verisign/storm/metrics/reporters/BaseKafkaReporterTest.java (hunk 5 above)
Comment (KevinJMao) [inline src/test/java/com/verisign/storm/metrics/reporters/BaseKafkaReporterTest.java]: The data provider feeds into both tests (for the `AvroKafkaReporter` and `SchemaRegistryKafkaReporter`).

**Pull request merged**